### DISCUSSION TITLE
Make Renovate bot to commit directly to main branch

### DISCUSSION THREAD

**Discussion by alvaroblazquez4567** [renovatebot/renovate#23324] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.120

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Cloud (bitbucket.org)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Let the Renovate bot to directly commit changes to the main branch.

### What you tried so far.

I have looked for this option in the configuration document:
https://docs.renovatebot.com/configuration-options/

I am aware about:

-  enable automerging. 
- avoid pr creation and merge the branch directly

What I want is to commit directly to main

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by HonkingGoose** [comment]
How about setting `automergeType=branch`? First read the docs though, to make sure it's what you want. 😉 Quote from [`automergeType` docs](https://docs.renovatebot.com/configuration-options/#automergetype):

> ## automergeType
> 
> This setting is only applicable if you opt in to configure `automerge` to `true` for any of your dependencies.
> 
> Automerging defaults to using Pull Requests (`automergeType="pr"`). In that case Renovate first creates a branch and associated Pull Request, and then automerges the PR on a subsequent run once it detects the PR's status checks are "green". If by the next run the PR is already behind the base branch it will be automatically rebased, because Renovate only automerges branches which are up-to-date and green. If Renovate is scheduled for hourly runs on the repository but commits are made every 15 minutes to the main branch, then an automerge like this will keep getting deferred with every rebase.
> 
> **Tip**
> If you have no tests but still want Renovate to automerge, you need to add `"ignoreTests": true` to your configuration.
> 
> If you prefer that Renovate more silently automerge _without_ Pull Requests at all, you can configure `"automergeType": "branch"`. In this case Renovate will:
> 
> - Create the branch, wait for test results
> - Rebase it any time it gets out of date with the base branch
> - Automerge the branch commit if it's: (a) up-to-date with the base branch, and (b) passing all tests
> - As a backup, raise a PR only if either: (a) tests fail, or (b) tests remain pending for too long (default: 24 hours)
> 
> The final value for `automergeType` is `"pr-comment"`, intended only for users who already have a "merge bot" such as [bors-ng](https://github.com/bors-ng/bors-ng) and want Renovate to _not_ actually automerge by itself and instead tell `bors-ng` to merge for it, by using a comment in the PR.
> If you're not already using `bors-ng` or similar, don't worry about this option.

**Reply by alvaroblazquez4567** [reply]
Thanks, but I was already aware of that option. While it may work almost the same, I was looking to avoid branch creation completely.

**Reply by HonkingGoose** [reply]
Hi @alvaroblazquez4567,

Can you explain _why_ you want Renovate to commit to your `main` branch directly? Are you doing something special with your Git repository that you do not want `renovate/` branches?

Most users would _want_ Renovate to run the test suite _first_ and only merge the branch when the tests pass. 🙃

**Reply by alvaroblazquez4567** [reply]
Sure @HonkingGoose. It is for projects I am applying trunk based development to. A dependency update would not be different to any other feature commit. All tests run before the code reaching prod.
It is not a blocker because those branches would be short lived, but they are of no use since CI is not triggered in branches at all.

I was just trying to set up Renovate in a full trunk based development environment.

I can only see that Renovate bot uses branches or prs (not sure there) as a way for keeping track dependency updates that have been already tried. So if a commit is reverted for a dependency update, Renovate bot will know about it and wont automerge that dependency version again. I dont know if that would be possible without branches and prs, just relaying on commits

**Comment by rarkins** [comment, marked as answer]
We don't plan to support direct committing, but if you have automergeType=branch and ignoreTests=true then you might it merging right away and achieving the same thing?

**Reply by alvaroblazquez4567** [reply]
Yes, I have considered that and it is a possible solution. I just wanted to check if avoiding branches was possible.
